src/app-layer-smtp.c

```diff
@@ -997,6 +997,10 @@ static int SMTPProcessReply(SMTPState *state, Flow *f, AppLayerParserState *psta
             state->parser_state |= SMTP_PARSER_STATE_COMMAND_DATA_MODE;
         } else {
             /* decoder event */
+            if (state->parser_state & SMTP_PARSER_STATE_PIPELINING_SERVER) {
+                // reset data mode if we had entered it prematurely
```

Comment:
Q: Why do we enter DATA mode before reply from the server?

Reply:
because the server advertised pipelining mode

Reply:
ok thanks.
Another question: Why don't we reset the data mode irrespective of pipelining mode if we're logging an event `SMTP_DECODER_EVENT_DATA_COMMAND_REJECTED`? The event name seems to suggest that DATA mode should not exist after this event.

Reply:
>  Why don't we reset the data mode irrespective of pipelining mode 

Because we only accepted it if we were in pipelining mode

Reply:
Indeed. I meant a patch like the following. Is this incorrect?

```
+            DEBUG_VALIDATE_BUG_ON(!(state->parser_state & SMTP_PARSER_STATE_PIPELINING_SERVER));
+            // reset data mode if we had entered it prematurely
+            state->parser_state &= ~SMTP_PARSER_STATE_COMMAND_DATA_MODE;
```

Reply:
This patch looks incorrect for the case the reply 354 for DATA happens without pipelining :
This would trigger the `DEBUG_VALIDATE_BUG_ON`

Reply:
ok. Misunderstood your last comment. Thank you!